### DISCUSSION TITLE
Extending myokit to 2d or 3d

### DISCUSSION THREAD

**Discussion by nmmhuynh** [MichaelClerx/myokit#1079] [Q&A] [answered]
Hi!
I've been trying to to use Myokit with our cardiac simulation codes in 2d and 3d, with diffusion.
It's almost clear to me how to extract arrays or variables of interest after a simulation run.
It is still a mystery to me how to (maybe better 'if it is possible to') give a variable as input.
For instance, I'd like to compute the transmembrane potential by solving a diffusion equation and then to use it as input to update the ionic model, in a time loop.
Or, in other words, I'd like to calculate the model derivatives, given a fixed voltage as an input as in :

   1.  Initialise system at t=0
    2. Get V from own codes
    3. Calculate derivatives and update state in Myokit -> get derivatives
    4. Increment time and go to step 2

I've read [the discussion here](https://github.com/myokit/myokit/discussions/1031) but since I'm a newbie with myokit, I could apply it to my situation.

Many thanks for any suggestion!!

**Comment by MichaelClerx** [comment, marked as answer]
Dear Mai,

That's an unusual request!
If you really wanted to, you could do something like:

    # Import myokit, create simulation.
    # This is slow so do it only once
    import myokit
    m = myokit.load_model('example')
    s = myokit.Simulation(m)

    # Initialise other simulation code here

    # Run simulation
    for t in ...:
        for cell in ...:
            # Get stimulus on/off
            pace = 1 if ... else 0
        
            # Optional: change model parameters
            # these will be updated internally and stay changed
            if cell in ...
                s.set_constant('ina.gNa', 0.1)
            else:
                s.set_constant('ina.gNa', 0.2)
            
            # Get expected net diffusion current (outwards = positive)
            i_diff = ...
            
            # Evaluate derivatives, without changing simulation state or
            s.evaluate_derivatives(
                inputs={'time': t, 'pace': pace, 'diffusion_current': i_diff},
                state=[-40, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 1e-4])

but this is assuming that your own diffusion codes are written in Python?

If all you need is some way to evaluate model equations based on e.g. CellML, it might be easier to use Myokit to just generate code for you? For example

    import myokit
    import myokit.formats
    m = myokit.load_model('example')

    # Generate runnable python code in the directory python-test
    e = myokit.formats.exporter('python')
    e.runnable('python-test', m)

    # Generate runnable matlab code in the directory matlab-test
    e = myokit.formats.exporter('matlab')
    e.runnable('matlab-test', m)

    # Generate runnable C/C++ code in the directory c-test
    e = myokit.formats.exporter('ansic')
    e.runnable('c-test', m)

    # Generate runnable cuda code in the directories cuda-test and cuda-rl-test
    e = myokit.formats.exporter('cuda-kernel')
    e.runnable('cuda-test', m)
    e = myokit.formats.exporter('cuda-kernel-rl')
    e.runnable('cuda-rl-test', m)

**Reply by nmmhuynh** [reply]
Dear Michael, 
thanks for your reply.
Yes, my own diffusion codes are written in Python and the first portion of script you shared is a good starting point.
However, I'd like to understand how to do the following: 
`       # Get expected net diffusion current (outwards = positive)
        i_diff = ...
        
        # Evaluate derivatives, without changing simulation state or
        s.evaluate_derivatives(
            inputs={'time': t, 'pace': pace, 'diffusion_current': i_diff},
            state=[-40, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 1e-4])

        # Use i_diff with own diffusion code for update potential V
        V = ....

         # Update state with newly computed V
         s.run( [time], [new V] )
`

At this stage, I would be happy to understand how to update the state with the newly computed V as initial condition.

**Reply by MichaelClerx** [reply]
You could use set_state to change the simulation state, then run, then repeat for every cell. But I'm struglling to see why you'd want to do that? A call to run() calls CVODE internally, which usually starts off with a relatively large number of evaluations to determine a good step size. Given that your diffusion model is usually the limiting factor for the step size, this would be highly inefficient?

You could try creating a multi-cellular simulation without diffusion, which will just use forward euler or rush-larsen under the hood.

But really if you're using you own diffusion model I don't think you're going to get any benefits from using Myokit over just writing the equations in plain Python. What am I missing here?

**Reply by MichaelClerx** [reply]
Here's a script to show you what I mean:

```python
import myokit
import numpy as np
import matplotlib.pyplot as plt

m, p, _ = myokit.load('example')
m.get('engine').add_variable('evaluations', rhs=0, binding='evaluations')
s = myokit.Simulation(m, p)

vs = ['engine.time', 'engine.evaluations', 'membrane.V']
d1 = s.run(1000, log=vs)
s.reset()

time = []
evals = []
vm = []

last_evals = 0
for i in range(1000):
    d2 = s.run(1, log=vs)

    # Evaluations are stored per call to run, so need to get cumulative evals
    # manually
    new_evals = np.asarray(d2['engine.evaluations']) + last_evals
    last_evals = new_evals[-1]
    evals.append(new_evals)
    time.append(d2['engine.time'])
    vm.append(d2['membrane.V'])

evals = np.concatenate(evals)
time = np.concatenate(time)
vm = np.concatenate(vm)


fig = plt.figure(figsize=(5, 5.5))
fig.subplots_adjust(0.16, 0.09, 0.99, 0.99, hspace=0.05)
grid = fig.add_gridspec(3, 1)

ax = fig.add_subplot(grid[0, 0])
ax.set_xticklabels([])
ax.set_ylabel('Vm (mV)')
ax.plot(d1.time(), d1['membrane.V'], label='One 1000ms run')
ax.plot(time, vm, ':', color='tab:orange', label='1000 1ms runs')
ax.legend(loc='upper right')

ax = fig.add_subplot(grid[1, 0])
ax.set_xticklabels([])
ax.set_ylabel('Evaluations (1 run)')
ax.plot(d1.time(), d1['engine.evaluations'])

ax = fig.add_subplot(grid[2, 0])
ax.set_xlabel('Time (ms)')
ax.set_ylabel('Evaluations (1000 runs)')
ax.plot(time, evals, color='tab:orange')

fig.align_labels()
plt.savefig('restarting.png')
```

![restarting](https://github.com/myokit/myokit/assets/517644/0b6f9623-292f-4390-b63b-b94c0164a24c)

We run twice: once with a single 1000ms call to CVODE, and once with 1000 1ms calls to CVODE. The top panel shows both APs are the same.

In the middle panel we show the number of evaluations for a single run. The evaluation count increases rapidly at the start (figuring out initial step size and stability information), the during the upstroke, then during repolarisation.

In the lower panel we show the cumulative number of evaluations when we split into 1000 1ms runs. Now the evaluations at the start of each call dominate, and we get a huge number of evaluations - so massively inefficient!

My guess is that this would outweigh the benefit of using myokit & cvode in your case
